src/Microsoft.Extensions.Configuration.AzureAppConfiguration/FailOverSupportedConfigurationClient.cs

```diff
@@ -0,0 +1,163 @@
+            this.failedRequestsToPrimaryConfigClient = 0;
+            this.clients = clients;
+        }
+
+        public Task<Response<ConfigurationSetting>> GetConfigurationSettingAsync(string key, string label = null, CancellationToken cancellationToken = default)
+        {
+            return ExecuteWithFailOverPolicyAsync(clients.Select((client) => new Func<Task<Response<ConfigurationSetting>>>(() => client.Client.GetConfigurationSettingAsync(key, label, cancellationToken))), cancellationToken);
+        }
+
+        public Task<Response<ConfigurationSetting>> GetConfigurationSettingAsync(ConfigurationSetting setting, bool onlyIfChanged = false, CancellationToken cancellationToken = default)
+        {
+            return ExecuteWithFailOverPolicyAsync(clients.Select((client) => new Func<Task<Response<ConfigurationSetting>>>(() => client.Client.GetConfigurationSettingAsync(setting,  onlyIfChanged, cancellationToken))), cancellationToken);
+        }
+
+        public AsyncPageable<ConfigurationSetting> GetConfigurationSettingsAsync(SettingSelector selector, CancellationToken cancellationToken = default)
+        {
+            return ExecuteWithFailOverPolicy(clients.Select((client) => new Func<AsyncPageable<ConfigurationSetting>>(() => client.Client.GetConfigurationSettingsAsync(selector, cancellationToken))), cancellationToken);
+        }
+
+        public void UpdateSyncToken(Uri endpoint, string syncToken)
+        {
+            this.clients.Single(client => client.Endpoint.Host.Equals(endpoint.Host)).Client.UpdateSyncToken(syncToken);
+        }
+
+        private async Task<Response<T>> ExecuteWithFailOverPolicyAsync<T>(IEnumerable<Func<Task<Response<T>>>> delegates, CancellationToken cancellationToken = default)
+        {
+            IList<Exception> exceptions = new List<Exception>();
+
+            for (var i = ShouldTryPrimaryConfigStore() ? 0 : 1; i < clients.Count(); i++)
+            {
+                var success = false;
+
+                try
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+                    var result = await delegates.ElementAt(i)();
+                    success = true;
+
+                    return result;
+                }
+                catch (RequestFailedException e) when (e.Status == (int)HttpStatusCode.ServiceUnavailable || e.Status == HttpStatusRequestThrottled)
+                {
+                    exceptions.Add(e);
+                    continue;
+                }
+                catch (Exception e)
+                {
+                    exceptions.Add(e);
+                    break;
+                }
+                finally
+                {
+                    if (i == 0)
+                    {
+                        UpdatePrimaryConfigStoreStatus(success);
+                    }
+                }
+            }
+
+            throw new AggregateException(exceptions);
```

Comment:
>throw new AggregateException(exceptions); [](http://example.com/codeflow?start=12&length=41)

The caller will now always get `AggregateException`, instead of the usual `RequestFailedException` or even `OperationCanceledException` that was expected earlier.

This has 2 side effects:
1. The public API `RefreshAsync` will stop throwing `RequestFailedException`, which users may be handling on their end. So this would be a breaking change.

2. We have special handling of `RequestFailedException` when the status is `HttpStatusCode.NotFound` since its not really an exception. It just means that a key-value does not exist, or was deleted from the store. 

If we do start throwing `AggregateException` with the new `FailOverSupportedConfigurationClient`, we need to handle this exception appropriately wherever we are using this client.

Reply:
Good catch! (Pun intended 🤣)

There is not a good way to fix this when we are trying the same request multiple times if the first fails.
Each request could fail with a different exception.

I have tried to set a logic which tries to address this problem.
1. The `OperationCanceledException` will be propagated to the caller as is.
2. If all of the exceptions in the `exceptions` list are `RequestFailedException`s, then we return the last `RequestFailedException`. (this will solve the problem when a kv doesn't exist in any of the replicas).
3. If there are multiple different types of exceptions, then we throw the `AggregateException` and let the caller handle these exceptions however they seem fit.

I believe this is a reasonable tradeoff between maintaining the backword compatibility and not hiding important information from exceptions when it is necessary.

I would like your thoughts on this.

Reply:
It seems this stems from the fact we're catching base `Exception`. We shouldn't be. We should only catch exceptions that we are aware of that can be triggered due to an App Configuration endpoint failing.

Reply:
I think we need 5 different catch blocks in this order:
1. Catch `RequestFailedException` when status is 503 or 429.
2. Catch `AggregateException` when all inner exceptions are `RequestFailedException` AND last inner `RequestFailedException` has status 503 or 429.
3. Catch any `RequestFailedException` --> using a fallback client won't fix this issue.
4. Catch `AggregateException` when all inner exceptions are `RequestFailedException` --> using a fallback client won't fix this issue.
5. Catch `OperationCanceledException` --> using a fallback client won't fix this issue.

Only the first two exceptions should trigger the fallback client.

Then the next question is - what should we throw when a fallback client is triggered and there is an exception? Should we throw an `AggregateException` of all exceptions that the current and previous clients got? Or should we just throw the most recent exception that the last client in the fallback clients' list got?

I think it makes sense to throw the very last exception received by the very last client. This is because the only way we get to the last client is if all the clients before it got either 503 or 429 exceptions. That way, we don't even need to change the calling code that's invoking this failover client's method.

Reply:
I am not sure why would we need to catch 3, 4 and 5.
We should propagate these exceptions to the caller 'as is' to maintain backwords compatibility.

Reply:
Ah right, so we can catch only the first two and bubble up everything else.

Reply:
Done!